com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
+                    continue;
+                }
+                if (supportsDropToFrame(thread, affectedFrame)) {
+                    popFrames.add(affectedFrame);
+                }
+            }
+        }
+        return popFrames;
+    }
+
+    /**
+     * Returns the stack frame that should be dropped to in the given thread after a
+     * hot code replace. This is calculated by determining if the threads contain
+     * stack frames that reside in one of the given replaced class names. If
+     * possible, only stack frames whose methods were directly affected (and not
+     * simply all frames in affected types) will be returned.
+     */
+    protected StackFrame getAffectedFrame(ThreadReference thread, List<String> replacedClassNames)
+            throws DebugException {
+        List<StackFrame> frames = getStackFrames(thread, false);
+        StackFrame affectedFrame = null;
+        for (int i = 0; i < frames.size(); i++) {
+            StackFrame frame = frames.get(i);
+            if (containsChangedType(frame, replacedClassNames)) {
+                if (supportsDropToFrame(thread, frame)) {
+                    affectedFrame = frame;
+                    break;
+                }
+                // The frame we wanted to drop to cannot be popped.
+                // Set the affected frame to the next lowest pop-able
+                // frame on the stack.
+                int j = i;
+                while (j > 0) {
+                    j--;
+                    frame = frames.get(j);
+                    if (supportsDropToFrame(thread, frame)) {
+                        affectedFrame = frame;
+                        break;
                     }
-                } catch (InterruptedException e) {
-                    logger.log(Level.INFO, "The hotcode replace completion was interrupted by " + e.getMessage(), e);
                 }
+                break;
             }
-            return new ArrayList<String>();
-        });
+        }
+        return affectedFrame;
+    }
+
+    protected boolean containsChangedType(StackFrame frame, List<String> replacedClassNames) throws DebugException {
+        String declaringTypeName = getDeclaringTypeName(frame);
+        // Check if the frame's declaring type was changed
+        if (replacedClassNames.contains(declaringTypeName)) {
+            return true;
+        }
+        // Check if one of the frame's declaring type's inner classes have
+        // changed
+        for (String className : replacedClassNames) {
+            int index = className.indexOf('$');
+            if (index > -1 && declaringTypeName.equals(className.substring(0, index))) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private String getDeclaringTypeName(StackFrame frame) throws DebugException {
+        return getGenericName(StackFrameUtility.getDeclaringType(frame));
+    }
+
+    private String getGenericName(ReferenceType type) throws DebugException {
```

Comment:
I remember @andxu implemented the similar format function in variable handler.  is there any possible to reuse?

Reply:
@andxu's change is not suitable here.  And he discarded his change.